### PR TITLE
[low-code connectors] Support for list-based stream slicers

### LINKED ISSUE #13002: Low-code connectors: Support custom stream slices
## What
The Metafields stream of the recharge connector reads records with hardcoded stream slices

```
    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
        owner_resources = ["customer", "store", "subscription"]
        for owner in owner_resources:
            yield from super().read_records(stream_slice={"owner_resource": owner}, **kwargs)
```

The streams for google-search-console also read records from a list of configurated values
```
    def stream_slices(
        self, sync_mode: SyncMode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
    ) -> Iterable[Optional[Mapping[str, Any]]]:
        for site_url in self._site_urls:
            yield {"site_url": site_url}
```

It should be possible to configure this list through the declarative DSL

## Connectors
- recharge
```
slicer:
  class_name: "airbyte_cdk.sources.declarative.stream_slicers.substream_slicer.ListIteratorSlicer"
  slice_values:
    - customer
    - store
    - subscription
  slice_definition:
    owner_resource: "{{ slice_value }}"
  state: "*ref(state)"
```
- github
refer to an array in the config
```
slicer:
  class_name: "airbyte_cdk.sources.declarative.stream_slicers.substream_slicer.ListIteratorSlicer"
  slice_values: "*ref(config['repositories'])
  slice_definition:
    repository: "{{ slice_value }}"
  state: "*ref(state)"
```
- google-search-console
Just like github. Worth pointing out this will also require exposing a method to sanitize the urls
```
slicer:
  class_name: "airbyte_cdk.sources.declarative.stream_slicers.substream_slicer.ListIteratorSlicer"
  slice_values: "*ref(config['site_url'])
  slice_definition:
    repository: "{{ slice_value }}"
  state: "*ref(state)"
```

### PR DESCRIPTION
**Pull request by girarda**
## What
* Add support for slicing the stream based off of the values in a list
* Closes https://github.com/airbytehq/airbyte/issues/13002

Sample config for recharge
```
slicer:
  class_name: "airbyte_cdk.sources.declarative.stream_slicers.substream_slicer.ListIteratorSlicer"
  slice_values:
    - customer
    - store
    - subscription
  slice_definition:
    owner_resource: "{{ slice_value }}"
```

## How
* Iterate over the values of the list and emit a stream slice for each value

## Recommended reading order
1. `airbyte-cdk/python/airbyte_cdk/sources/declarative/stream_slicers/list_stream_slicer.py`
2. `airbyte-cdk/python/unit_tests/sources/declarative/stream_slicers/test_list_slicer.py`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-cdk/python/unit_tests/sources/declarative/stream_slicers/test_list_slicer.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    [
+        (
+            "test_single_element",
+            ["customer", "store", "subscription"],
+            {"owner_resource": "{{ slice_value }}"},
+            [{"owner_resource": "customer"}, {"owner_resource": "store"}, {"owner_resource": "subscription"}],
+        ),
+    ],
+)
+def test_substream_slicer(test_name, slice_values, slice_definition, expected_slices):
```
Comment (sherifnada) [inline airbyte-cdk/python/unit_tests/sources/declarative/stream_slicers/test_list_slicer.py:21]: is it useful to also have a test for the case where the list comes from the config like github?

**Review by sherifnada — APPROVED** [review]
LGTM but I have one question: what is the purpose of the `state: *ref(state)` arg? what does it achieve? i'm not seeing it used in the slicer

**Comment by girarda** [comment]
> LGTM but I have one question: what is the purpose of the `state: *ref(state)` arg? what does it achieve? i'm not seeing it used in the slicer

deleted from the sample. other stream slicers such as the DatetimeStreamSlicer also do not need the state.

File: airbyte-cdk/python/unit_tests/sources/declarative/stream_slicers/test_list_slicer.py (hunk 1 above)
Comment (girarda) [reply airbyte-cdk/python/unit_tests/sources/declarative/stream_slicers/test_list_slicer.py]: yes, done

**Pull request merged**